provider/google: Created atomic operation for restoring snapshot, implemented job executor to run terraform, added logic for importing terraform state files

nwwebb opened the pull request:
Added a job executor mostly copied from the one @duftler [wrote for rosco](https://github.com/spinnaker/rosco/tree/f165221be7bece9492af99f8501b154bc168cfff/rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/jobs). This required me to add two dependencies to clouddriver (commonsExec and rxJava). I also created an atomic operation for restoring snapshots. Right now the operation successfully imports the state of existing resources into a terraform state file, but only works locally if terraform v7.1 or greater is found in PATH. Also environment variables must be set for terraform to [properly find google credentials](https://www.terraform.io/docs/providers/google/). PTAL @lwander

lwander commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/jobs/local/JobExecutorLocal.groovy:
```diff
@@ -0,0 +1,172 @@
+      new Action0() {
+        @Override
+        public void call() {
+          ByteArrayOutputStream stdOutAndErr = new ByteArrayOutputStream()
```
Where is this set to read from both stdout & stderr?

nwwebb replied:
when only one argument is specified for [PumpStreamHandler](https://commons.apache.org/proper/commons-exec/apidocs/org/apache/commons/exec/PumpStreamHandler.html) it is set for stdout and stderr

lwander commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/jobs/local/JobExecutorLocal.groovy:
```diff
@@ -0,0 +1,172 @@
+
+            commandLine.addArguments(arguments, false)
+          } else {
+            log.info("No tokenizedCommand specified for $jobId.")
```
You don't need to log this here, since it the exception below will make it into the logs.

lwander commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/jobs/local/JobExecutorLocal.groovy:
```diff
@@ -0,0 +1,172 @@
+              // If a watchdog is passed in, this was an actual time-out. Otherwise, it is likely
+              // the result of calling watchdog.destroyProcess().
+              if (w) {
+                log.info("Job $jobId timed-out (after $timeoutMinutes minutes).")
```
I think this should be a `warn`, not an `info`.

lwander commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/jobs/local/JobExecutorLocal.groovy:
```diff
@@ -0,0 +1,172 @@
+            jobStatus.state = JobStatus.State.COMPLETED
+            jobStatus.result = JobStatus.Result.SUCCESS
+          } else {
+            jobStatus.state = JobStatus.State.CANCELED
```
Why `CANCELED`? I think `130` is the de-facto `ctrl-c` exit code, but a failure doesn't usually imply a cancellation.

lwander commented on clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/RestoreSnapshotAtomicOperation.groovy:
```diff
@@ -0,0 +1,149 @@
+    this.project = credentials.project
+
+    // Remove the terraform state file if it exists
+    def command = ["rm", "terraform.tfstate"]
```
I think these statefiles shouldn't all have the same name, as the executions can happen in parallel.

nwwebb commented:
addressed comments @lwander. Also right now the timeout is specified by jobs.local.timeoutMinutes in the yml file and defaults to 10 minutes. I took this directly from rosco, so let me know if this seems reasonable for terraform

lwander commented:
LGTM, any comments, @duftler?

duftler commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/jobs/config/LocalJobConfig.groovy:
```diff
@@ -0,0 +1,36 @@
+    new JobExecutorLocal()
+  }
+}
+
```
Extraneous blank line.

duftler commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.groovy:
```diff
@@ -67,4 +67,5 @@ final class AtomicOperations {
 
   // Serialize operations
   public static final String SERIALIZE_APPLICATION = "serializeApplication"
```
This line isn't touched in this pr, but hadn't we discussed changing `serializeApplication` to `captureSnapshot` or something like that?

nwwebb replied:
I am planning on changing the names for everything in a separate pr

duftler replied:
sgtm

duftler commented on clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/converters/SerializeApplicationAtomicOperationConverter/RestoreSnapshotAtomicOperationConverter.groovy:
```diff
@@ -0,0 +1,42 @@
+  RestoreSnapshotDescription convertDescription(Map input) {
+    GoogleAtomicOperationConverterHelper.convertDescription(input, this, RestoreSnapshotDescription)
+  }
+
```
Extraneous blank line.

duftler commented on clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/description/SerializeApplicationDescription/RestoreSnapshotDescription.groovy:
```diff
@@ -0,0 +1,25 @@
+class RestoreSnapshotDescription extends AbstractGoogleCredentialsDescription {
+  String applicationName
+  String accountName
+  Long snapshotTimestamp //timestamp of the snapshot to restore
```
Does this comment really add information that isn't contained in the field name itself?

duftler replied:
I might be missing it, but I don't see where `snapshotTimestamp` is used here? From the sample `curl` command on the operation, it looks optional; but what is done with it if it is specified?

nwwebb replied:
this pr doesn't include pulling the snapshot from front50 and restoring it, I am working on that now

duftler replied:
sgtm

duftler commented on clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/RestoreSnapshotAtomicOperation.groovy:
```diff
@@ -0,0 +1,165 @@
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperation
+import com.netflix.spinnaker.clouddriver.security.AccountCredentialsRepository
+import org.springframework.beans.factory.annotation.Autowired
+
```
Extraneous blank line.

duftler commented on clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/RestoreSnapshotAtomicOperation.groovy:
```diff
@@ -0,0 +1,165 @@
+      throw new IllegalArgumentException(jobStatus.logsContent)
+    }
+  }
+
```
2 extraneous blank lines.

duftler commented on clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/RestoreSnapshotAtomicOperation.groovy:
```diff
@@ -0,0 +1,165 @@
+    this.project = credentials.project
+
+    // Make directory for terraform files
+    def makeDirCommand = ["mkdir", "$applicationName-$accountName"]
```
Would prefer to use the java-provided mechanism for creating directories as opposed to relying on the job runner mechanism.
Same for deleting the directory, even though it will require a bit of work to first empty the directory.

nwwebb commented:
address your comments @duftler

duftler commented:
LGTM.
